File: 313.超级丑数.cpp
/*
 * @lc app=leetcode.cn id=313 lang=cpp
 *
 * [313] 超级丑数
 */

// @lc code=start
class Solution
{
public:
    int nthSuperUglyNumber(int n, vector<int> &primes)
    {
        vector<int> results;
        int k = primes.size();
        vector<int> resultIndex(k, 0);
        results.push_back(1);
        for (int i = 1; i < n; i++)
        {
            int ugly = INT_MAX;
            for (int j = 0; j < k; j++)
                ugly = min(ugly, primes[j] * results[resultIndex[j]]);
            results.push_back(ugly);
            for (int j = 0; j < k; j++)
                if (ugly == primes[j] * results[resultIndex[j]])
                    resultIndex[j]++;
        }
        return results[n - 1];
    }
};
// @lc code=end

File: 313.超级丑数.py
#
# @lc app=leetcode.cn id=313 lang=python3
#
# [313] 超级丑数
#

# @lc code=start
class Solution:
    def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
        k = len(primes)
        results = [1, ]
        resultIndex = [0]*k
        for i in range(1, n):
            ugly = min([primes[j]*results[resultIndex[j]] for j in range(k)])
            results.append(ugly)
            for j in range(k):
                if ugly == primes[j]*results[resultIndex[j]]:
                    resultIndex[j] += 1
        return results[-1]
# @lc code=end
